mbed_lstools/lstools_base.py

```diff
@@ -176,37 +179,109 @@ def _fs_before_id_check(self, device, filter_function, read_details_txt):
-                device.update({"daplink_%s" % f.lower().replace(' ', '_'): v
-                               for f, v in details_txt.items()})
-
+            self._update_device_from_fs(device, read_details_txt)
             return device
         else:
             return None
 
+    def _update_device_from_fs(self, device, read_details_txt):
+        """ Updates the device information based on files from its 'mount_point'
+            @param device Dictionary containing device information
+            @param read_details_txt A boolean controlling the presense of the
+              output dict attributes read from other files present on the 'mount_point'
+        """
+        if not device.get('mount_point', None):
+            device['device_type'] = 'unknown'
+            return
+
+        device['device_type'] = self._detect_device_type(device['mount_point'])
+        device['target_id'] = device['target_id_usb_id']
+
+        {
+            'daplink': self._update_device_details_daplink,
+            'jlink': self._update_device_details_jlink
+        }[device['device_type']](device, read_details_txt)
+
+    def _detect_device_type(self, mount_point):
+        """ Returns a string of the device type
+            @return 'daplink' or 'jlink'
+        """
+
+        files = [f.lower() for f in os.listdir(mount_point)]
+        return 'jlink' if 'segger.html' in files else 'daplink'
+
+
+    def _update_device_details_daplink(self, device, read_details_txt):
+        """ Updates the daplink-specific device information based on files from its 'mount_point'
+            @param device Dictionary containing device information
+            @param read_details_txt A boolean controlling the presense of the
+              output dict attributes read from other files present on the 'mount_point'
+        """
+        self._update_device_from_htm(device)
+        if read_details_txt:
+            details_txt = self._details_txt(device['mount_point']) or {}
+            device.update({"daplink_%s" % f.lower().replace(' ', '_'): v
+                           for f, v in details_txt.items()})
+
+        if device['target_id']:
+            platform_data = self.plat_db.get(device['target_id'][0:4],
+                                             device_type='daplink',
+                                             verbose_data=True)
+            device.update(platform_data or {'platform_name': None})
+        else:
+            device['platform_name'] = None
+
+    def _update_device_details_jlink(self, device, read_details_txt):
+        """ Updates the jlink-specific device information based on files from its 'mount_point'
+            @param device Dictionary containing device information
+            @param read_details_txt A boolean controlling the presense of the
+              output dict attributes read from other files present on the 'mount_point'
+        """
+        files = os.listdir(device['mount_point'])
```

Comment:
Could we list the directory only once some how? We could do the listdir outside this function and `_detect_device_type` and pass it into both. I think that this may have an impact on performance (we're somewhat performance critical to RAAS, IIRC) and we want to avoid making things any slower than they have to be.